### ISSUE TITLE
[URL Unfurling] the need for async event-driven backend

### ISSUE THREAD

**Issue by igor-sirotin** [status-im/status-go#4196]
# Description

So I'm going to speak about this design:

<img width="1105" alt="image" src="https://github.com/status-im/status-desktop/assets/25482501/b0ceba97-5eff-47c0-b4c6-cc00fc34be4e">

We have a few design features to keep in mind here:
1. About the settings
    1. `UnfurlingMode` setting (can be `EnableAll`, `DisableAll`, `AlwaysAsk` and in post-mvp `SpecifyBySite`).
    2. `AlwaysAsk` can be proceeded with `Allow for this message`
    3. Status links should always be unfurled, no matter of the settings. Because we consider it as private and secure.
2. About the unfurling state. It can be on of:
    1. Successfully unfurled
    2. Unfurling in progress
    3. Able to unfurl this URL, but failed unfurling it (e.g. because of internet connection loss)
    4. Not able to unfurl this URL (e.g. a GIF link)
    5. Unfurling of this URL is disabled by settings (e.g. `mode == DisabledAll`, or `mode == SpecifyBySite` and it's in blocked)
    6. Unfurling of this URL is not YET allowed (e.g. `mode == AlwaysAsk`, or `mode == SpecifyBySite` and it's a new site)
3. About separating fields of responsibility between app and status-go:
    1. It better to implement as much as possible in status-go
    2. status-go knows most of the needed information (e.g. has access to settings)
    3. the only thing that status-go doesn't know about is "current message". Which is needed to implement the "Allow unfurling for this message" feature\*. 

\* "Allow unfurling for this message"
This setting is "message-persistent". So:
- start typing a message for user A, 
- allow unfurling for this message
- switch to chat with user B
- disable unfurling for this message
- switch back to chat with user A → unfurling is still enabled in this message.
- clear the message, start typing a new one → unfurling is again not enabled for this message.

status-go knows nothing about those chat switches. And because of this we now have the implementation of most of the logic in the App.

# The situation

To support all of this, and because of points 3.1 and 3.3, we will have to implement a lot of logic in the desktop app (and duplicate it in mobile app).

1. we wanted to make the settings handling in the status-go, but we can't fully implement it there, because it knows nothing about "current message".
2. we want to always unfurl Status links. So we will need to filter them on the desktop app, no matter of the current settings, and unfurl. But keep out all other links meanwhile.
3. we want to skip GIF links, so we're filtering them on the desktop side as well.

This will become a very messy implementation.

# Proposal

**Implement everything in status-go**

For this to be possible we will need an "event-driven" backend implementation with a `session-id` conecpt.

### What's "event-driven"?

Unfurl all URLs in parallel.
When the unfurling is not supported for given link, immediately send event with "UnsupportedURL" state.
When the unfurling is not yet allowed given link, send an `AskUserState` event, so that app could show the settings card in the chat composer.
When unfurling is supported and started, send `LoadingState` event.
And finally when the unfurling finished/failed, send a corresponding event to the app.

### Why "event-driven"?

First of all, because it will allow to update the UI link preview card as soon as it's ready. Now `UnfurlURLs` endpoint receives an list of urls and you won't get the result until all of unfurling finishes. Even if some of the links we don't even support (e.g. gif).

Mentioning the unsupported links, we shouldn't even show a "loading" animation for them. There shouldn't be a card for those at all. With current endpoint it's impossible to implement.

### What's a `session-id`?

This is almost a "message-id" actually, but for not-yet-sent messages. Maybe a "message-draft-id" also works. But think of it as  of a Session.

1. Sessions are stored in status-go with we provide `Start`/`Stop` API.
2. Starting/Stopping
    1. A session is started by app when a new message is started to type.
    2. A session is closed by app when the message is cleared or sent.
3. For each session, we save a "session-persistent-unfurling-mode".
We currently have this is Nim. It's mode that allows to "enable unfurling for this message".
4. We provide the `session-id` to the `UnfurlURLsAsync` endpoint.
With this it will be able to unfurl the URLs with full respect to settings, even for particular message.

## More benefits 🎉 

1. I think it should be possible to implement the unfurling in a single status-go call.
Currently we call `GetTextURLs` and then `UnfurlURLs`. Which makes sense ofc. We use it to first get the list of URLs, draw the cards with loading animation and then start unfurling.
But it we will get `Loading` state from status-go, then we can make it in a single call.

2. We will be able to move the link previews cache to status-go as well.
With a single-call to the backend, we will keep sending the message text to the backend during typing, as we don now.
But the backend could also track which URLs were already unfurled for this session. And if the link was already unfurled, just send no event. That's very simple 🙂

**Comment by igor-sirotin** [comment]
cc @alexjba @ilmotta @jrainville

**Comment by alexjba** [comment]
@igor-sirotin Nice work on this analysis!

> I think it should be possible to implement the unfurling in a single status-go call.
Currently we call GetTextURLs and then UnfurlURLs. Which makes sense ofc. We use it to first get the list of URLs, draw the cards with loading animation and then start unfurling.
But it we will get Loading state from status-go, then we can make it in a single call.

We're currently using `GetTextURLs` as an hyperlink detector to format the text in the input. This API is really convenient as we don't have to duplicate the logic in the application. If we're going for a signals based implementation, I think we should either keep this API or create a new event `urlsDetected`. I'd keep this API because it's more versatile and it could replace the regular expressions we're using in other places.

**Comment by jrainville** [comment]
Before giving my full opinion on this, I need a clarification about this point:

> * "Allow unfurling for this message"
This setting is "message-persistent"

Does this mean all of those are per message, or there is a sub-setting that makes it per message? I didn't see the design, so I don't know fully how it works. If you could link the design for me, that'd be nice.

Anyway, IMO, all those site settings (whitelist-ish) should be a global one. It makes it much simpler because that way we do not have to have the logic in the app and instead only in status-go, like you pointed out.

Maybe I'm misunderstanding, but I feel like this is something we could get the designers to drop.

**Comment by igor-sirotin** [comment]
@jrainville 

Sorry, I should've linked the design of course: https://www.figma.com/file/Mr3rqxxgKJ2zMQ06UAKiWL/%F0%9F%92%AC-Chat%E2%8E%9CDesktop?type=design&node-id=23648-222385&mode=design&t=aJwzjvhn8OrdO47Z-4

So we have this setting `AlwaysAsk`. Which means that for each URL you write in the message you will get an "asking card" if this URL should be unfurled. An for this case there's a `Show for this message` options, which temporarily allows the unfurling of all links for this particular draft message.

This `Enable unfurling for this message` makes sense to me, because the `UnfurlURLs` endpoint will be called on any link change. Yes, the preview is cached on the Nim side, but if you edit the link or paste another one - you will get this annoying "Asking card".

We can ofc ask designers to remove this feature, but it doesn't remove the need for async event-driven API. There're other pros of this too 🙂

**Comment by jrainville** [comment]
Alright, I see nothing wrong with your proposal so far. I'll leave it up to you guys to decide what is the best for this.

With the design, I can see why the "for this message only" unfurl is kinda useful, though it's not necessary, since users can always just click the X anyway. So I'll leave it up to you to decide if you think implementing it is worth it in terms of time and work. You can always start with the global settings and then do the per message setting.

Anyway, like you said, you still need the async event driven API.

**Comment by ilmotta** [comment]
Hey @igor-sirotin, thank you for the detailed analysis.

I have a few concerns to share, but I couldn't keep the text shorter 🤷🏼 Please, correct me if I made any sort mistake in my analasys.

> Unfurl all URLs in parallel.

Do you mean like unfurling them concurrently? This can be accomplished in the current implementation just as well. Also related, we currently timeout the HTTP requests at 15s, which is too high. Ideally the requests should use an exponential backoff strategy with a short timeout.

Just a reminder that we agreed we wouldn't unfurl more than like 5 URLs for each message the user sents. I think this limit is not yet set on status-go, right? For the mobile client it's important to prevent users from unfurling too many URLs at once for performance reasons. This setting is also important to prevent big waku messages while we don't have a better mechanism in place. Maybe the limit should be enforced at the client side because the desktop client has more resources available and could probably handle more unfurled data than mobile.

> Mentioning the unsupported links, we shouldn't even show a "loading" animation for them. There shouldn't be a card for those at all. With current endpoint it's impossible to implement.

The URL for a GIF, for example, still needs to be fetched because status-go doesn't know the mime type yet, and this can take any amount of time, so what do you mean by "we shouldn't even show the loading animation"? We could optimize this process a little bit by using a blocklist of hostnames and/or parts of the URL path and skip them as NOPs if we know they will return unsupported resources to status-go.

For example, the current status-go code has an optimization to check if the URL has one of the supported image extensions, but we never check the opposite, that the URL contains an unsupported extension, and thus should be skipped immediately.

> I think it should be possible to implement the unfurling in a single status-go call.

Yes, and I believe this is always possible for any feature really. It's just a matter of moving state to the backend and keeping it in sync between client & server, but this idea can get ugly very quickly. State synchronization is the root of a lot of evil. So to me it's about finding a balance. Also, the DX is painful when implementing features full stack as you need to waste a lot of time on slow status-go compile cycles and debugging two different systems in coordination. In my experience so far, I think backend systems should be coupled with the minimum amount of client state, unless of course there's a good reason.

> But the backend could also track which URLs were already unfurled for this session. And if the link was already unfurled, just send no event. That's very simple

I may be missing something, but I think the cache server (status-go here) must somehow report if it's a cache hit or miss and provide the cached data if we want to remove the cache completely from the client side.

Let's analyze this scenario:

1. User types URL `A`.
2. An RPC request is sent to status-go, status-go unfurls `A`, caches it and soon publishes the event.
3. The user sees `A` unfurled on screen.
4. The user then replaces the text with URL `B`.
5. An RPC request is sent to status-go, status-go unfurls `B`, caches it and soon publishes the event.
6. So far so good, the user sees `B` unfurled on screen. At this point, status-go has cached `A` and `B` for the current *session*.
7. The user replaces `B` with `A` again (it was a mistake to type `B` :shrug:) An RPC request is sent requesting `A` to be unfurled.
8. status-go finds that `A` is cached and does nothing.
9. The client doesn't know what to do with no event & no cached data, so the user won't see `A` unfurled.

Could you explain in more detail how it would work for the described scenario if the client won't store the cached data?

> We will be able to move the link previews cache to status-go as well. With a single-call to the backend, we will keep sending the message text to the backend during typing, as we don now. But the backend could also track which URLs were already unfurled for this session. And if the link was already unfurled, just send no event. That's very simple 🙂

I disagree in general about the sentiment that this issue's proposal is simple or simpler.

In terms of raw performance, the original code design for link previews preferred the cache on the client because that means the client would be able to much more quickly check if the URL had been unfurled than going through the whole RPC lifecycle. But I designed it with the mobile restrictions in mind. It's a different for desktop. With the new proposal, every time the user finishes typing (in mobile we debounce at 400ms), a full RPC request/response cycle will need to be performed because the client won't be able to know if the URL is cached. This is not optimal for mobile in the chat screen, which is quite heavy already.

Also, the caching logic takes no more than ~12 lines of Clojure, and it's quite flexible. For example, it's not clear for the mobile environment how aggressively we should cache unfurled data. This flexibility is important at this stage of the product. Currently we clear the cache when the user switches to another chat, but this is not optimal IMO. By moving this logic to status-go, we'll probably lose this flexibility since desktop and mobile have different performance characteristics, so we'll need to settle for one side or the other, or make the cache more configurable, which is more complex.

Currently, unfurling is mostly stateless, old-fashion request-response cycle, all or nothing. With the proposal, the client will need to process signals for the unfurling state changes. This will incur in more overhead and more code for mobile (can't say much about desktop). Also, the client will need to keep the backend state synchronized, which is yet more code. It's not really going to be a single call to status-go :)

About the signals published by status-go, we'll need to deal with the problem of signals arriving out of order. For instance, for the current mobile implementation, when the input text is sent to status-go to be unfurled, the unfurled response is discarded if the request ID is not the same as the originating request. This prevents a type of bug that happened with the mentions system recently. For example, when the user types a URL `A` that takes longer to unfurl than the replacement URL `B`, when `A` response arrives it should not replace `B` because `B` is the last call. With the proposal relying on signals this will be more tricky to solve.

Example scenario:

1. The user types URLs `A` and `B`.
2. status-go will publish two signals for `A` and `B` once they are unfurled, but while status-go was unfurling `A` and `B`, the user changed the input text and now the input text has URLs `A` and `C`. This is easy to happen because the event is debounced to higher values in mobile.
3. The signals for `A` and `B` arrive before the next RPC call for `A` and `C` arrive at status-go. But the input text doesn't have `B` anymore, so the client will need to ignore unfurled data for `B`.

Therefore, I believe the client will still need to keep track of which URLs are currently being unfurled and have some logic to ignore events for URLs that are no longer in the input text to sort of implement this _last request wins_ semantic. If I'm correct in my analysis of this scenario, then it's another example that it won't be as simple as just calling a single endpoint and letting status-go do all the rest.

--------

Having said all of that 🙉 my overall view of the proposal is that it will allow us to solve some problems better. All of this comes at the cost of complexity and higher likelihood for bugs, which was the reason I avoided this kind of solution in the beginning. But maybe the time has come for it? I've almost never seen evented code be simpler. Additionally, I think the majority of the time, users don't type/paste more than 1 or 2 URLs, so in a way, this issue is trying to solve 20% of the curve with a more complex, general solution. Simply improving how the HTTP calls are dispatched should be more than enough to unfurl most 1-2 URLs in less than 3-5s, which I think is pretty decent. Or is there a killer business requirement that forces us to implement the proposal (sorry if I missed this)?

As the product evolves and gets real users, I can imagine we might settle to a more complex solution to cater for smaller details, but right now I think a stateless approach to unfurl URLs is fine.

Side note/opinion: I don't agree, nor like the amount of complexity link previews have due to the requirement that there can be `N` URLs to preview, both in designs and code. If this requirement was removed I think we would still make our future users very happy for a long time, just as they are in other messaging apps.

I'm not convinced this is the technical direction to go right now, but if that's something you want to pursue I won't push back ;)

**Comment by igor-sirotin** [comment]
@ilmotta thanks for your reply and reasonable counter-arguments!

> I disagree in general about the sentiment that this issue's proposal is simple or simpler.

👍 👍 👍 

I think I underestimated the difficulties of this idea. But that's exactly why we're here all together 😄 

Though I still think it's possible to implement, I see that it's more difficult than I thought. But keeping it as is also not a good idea. I'm thinking now that maybe in `UnfurlURLs` we shouldn't process the settings at all? Or make another endpoint to be called between `GetTextURLs` and `UnfurlURLs`, smth like `GetSupportedURLs`, as @alexjba proposed.

Let's discuss tomorrow!

Below I answered your questions anyway, not to leave them hanging in the air.

_______

> > Unfurl all URLs in parallel.
>
> Do you mean like unfurling them concurrently? This can be accomplished in the current implementation

Sure, I understand it can be achieved with current API.
But I meant not just the concurrency, but that the results will also concurrently arrive to the client as soon as they're ready.

> ... we wouldn't unfurl more than like 5 URLs

Indeed, I remember we talked about this! We should probably force this to be implemented.

> > Mentioning the unsupported links, we shouldn't even show a "loading" animation for them.
>
> The URL for a GIF, for example, still needs to be fetched

In current implementation we filter `*.gif` urls on the desktop client before sending them to `UnfurlURLs`, because as agreed they're unfurled on the receiver side.

And I think for those links we don't want to show the loading.

> Also, the DX is painful when implementing features full stack as you need to waste a lot of time on slow status-go compile cycles and debugging two different systems in coordination. In my experience so far, I think backend systems should be coupled with the minimum amount of client state, unless of course there's a good reason.

I tend to implement more in status-go, because it's easier to test and maintain. And any applied fix goes straight to both clients.

Maybe it's not the case for status-mobile, but at least with status-desktop, we don't have debugging ability, and we don't have unit tests, only UI e2e tests. So it's much easier to develop things in status-go with tests 🙂 


> > But the backend could also track which URLs were already unfurled for this session.
>
> Could you explain in more detail how it would work for the described scenario if the client won't store the cached data?

yeah, I see... I guess it won't work that simple 😄 

I guess we could fix this this way: status-go should return the cached data IF both applies:
1. it exists in the cache
2. it wasn't requested in last `UnfurlURLsAsync` call

But I didn't think about this well and pretty sure it won't work propertly in all cases.

And you're definitely right that it's gonna be a pain to sync all the events.

> Therefore, I believe the client will still need to keep track of which URLs are currently being unfurled and have some logic to ignore events for URLs that are no longer in the input text to sort of implement this last request wins semantic. If I'm correct in my analysis of this scenario, then it's another example that it won't be as simple as just calling a single endpoint and letting status-go do all the rest.

I think `RequestID` should help here. 
1. client only accepts events with latest `RequestID`
2. (maybe) status-go stops previous requests, if newer is received. 
Event better if stopping can be smart. E.g. keep loading and cache results, but don't fire events at the end. Or keep unfurling the URLs that are also found in the new request.
